### ISSUE TITLE
Freeze with ATP, Macbook Pro 2,1

### ISSUE THREAD

**Issue by KevinMcA** [rgrover/freebsd-atp#1]
I see that this was reported earlier this year, but the person who started the PR never followed up on the troubleshooting.

Symptom: After a few minutes, atp driver on older Macbook (vendor=0x05ac product=0x021a devclass=0x00 devsubclass=0x00 sernum="" release=0x0018 mode=host intclass=0x03 intsubclass=0x01 intprotocol=0x02) will become completely unresponsive. The issue appears to be calibration drift due to thermals in the Macbook. We all know the touchpads can drift significantly due to temperature changes and it would appear that the touchpad is not being recalibrated, thus causing the issue.

What is interesting is that as it becomes unresponsive, it appears as though the top of the touchpad fades first and slowly becomes unresponsive vertically, losing data output until the area of response finally reaches the bottom of the trackpad. Once the bottom goes away, there's no response at all so the miscalibration appears to be on the X axis over just a few minutes worth of temperature change.

Replacing the atp.c with the following date information:
__FBSDID("$FreeBSD: releng/10.1/sys/dev/usb/input/atp.c 263063 2014-03-12 07:15:41Z hselasky $");

With an earlier version from FreeBSD 10.0 release:
__FBSDID("$FreeBSD: release/10.0.0/sys/dev/usb/input/atp.c 233774 2012-04-02 10:50:42Z hselasky $");

... resolves the issue completely, but of course lacks all of the new, desired features. Given that there is such a great difference in code between the two, I don't know where to start troubleshooting but will be happy to assist.

Hope this helps to give you an idea of where the issue might be, saw that previous person's information and hope this helps to clarify a starting point. There's still a lot of these old Macs out there and would love to resolve this.

Thanks!

**Comment by rgrover** [comment]
Hi,
Thanks for reporting this. My old Mac is covered in dust, but I'll try to revive it. If I can't then I'd appreciate some help with proxy debugging.

thanks.

**Comment by KevinMcA** [comment]
Be happy to help, will need to do kernel compiles to test, that usually kills a day. (grin)

Shall I set up anything for debugging? I normally build with all debugging disabled since my kernels have been boring for a while. But yeah, it's clearly some sort of reset not happening to reposition x=1, y=1 (mostly x axis) for the 0,0 position on the trackpad. I expect that once we can get it to find 0,0, it'll be quickly resolved. It was easy to find the reset in the old code, not sure now that it's mixed with the Wellspring stuff that was added. I suspect that's where it might have gone off the tracks.

**Comment by rgrover** [comment]
give me until the weekend to dust off my old mac and to see off the guests who're visiting at the moment. :) I'll get back to you by the weekend.

**Comment by KevinMcA** [comment]
My bad ... long night. I meant the "Y axis" is what's going off. I haven't noticed it being affected in the horizontal detection. It's purely a vertical thing.

**Comment by KevinMcA** [comment]
No worries! I know how that is. There are no fire alarms ringing, so we're cool. :)

**Comment by KevinMcA** [comment]
Just wanted to add a little more info since I just collected it for when there's time. In usb/188829, HPS had asked for some usbdump information, so dropping that in here for you, it wasn't useful. I will also include the xorg info on the device as detected. I still believe that something that should be resetting to zero is incrementing in an unplanned fashion on the Y axis for idle position, still looking to see if I might spot it, both times after a cool down, it took less than 3 minutes to vanish off the bottom with no further action. Even the button was unresponsive:

# usbdump -i usbus0 -f 3 -s 65535 vvv

^C^C^C^C^C

0 packets captured
7120 packets received by filter
0 packets dropped by kernel

------------ XORG info --------------

[    87.849](II) Using input driver 'kbd' for 'product 0x1000'
[    87.849](**) product 0x1000: always reports core events
[    87.849](**) product 0x1000: always reports core events
[    87.849](**) Option "Protocol" "standard"
[    87.849](WW) Option "Device" requires an string value
[    87.849](**) Option "XkbRules" "base"
[    87.849](**) Option "XkbModel" "pc105"
[    87.849](**) Option "XkbLayout" "us"
[    87.849](**) Option "config_info" "hal:/org/freedesktop/Hal/devices/usb_device_5ac_1000_noserial_if0"
[    87.849](II) XINPUT: Adding extended input device "product 0x1000" (type: KEYBOARD, id 6)
[    87.859](II) config/hal: Adding input device Apple Internal Keyboard  Trackpad
[    87.859](II) LoadModule: "mouse"
[    87.860](II) Loading /usr/local/lib/xorg/modules/input/mouse_drv.so
[    87.873](II) Module mouse: vendor="X.Org Foundation"
[    87.873]    compiled for 1.12.4, module version = 1.9.0
[    87.873]    Module class: X.Org XInput Driver
[    87.873]    ABI class: X.Org XInput driver, version 16.0
[    87.873](II) Using input driver 'mouse' for 'Apple Internal Keyboard  Trackpad'
[    87.873](**) Apple Internal Keyboard  Trackpad: always reports core events
[    87.873](**) Option "Device" "/dev/sysmouse"
[    87.873](==) Apple Internal Keyboard  Trackpad: Protocol: "Auto"
[    87.873](**) Apple Internal Keyboard  Trackpad: always reports core events
[    87.873](==) Apple Internal Keyboard  Trackpad: Emulate3Buttons, Emulate3Timeout: 50
[    87.873](**) Apple Internal Keyboard  Trackpad: ZAxisMapping: buttons 4 and 5
[    87.873](**) Apple Internal Keyboard  Trackpad: Buttons: 5
[    87.873](**) Option "config_info" "hal:/org/freedesktop/Hal/devices/usb_device_5ac_21a_noserial_if1"
[    87.873](II) XINPUT: Adding extended input device "Apple Internal Keyboard  Trackpad" (type: MOUSE, id 7)
[    87.874](**) Apple Internal Keyboard  Trackpad: (accel) keeping acceleration scheme 1
[    87.874](**) Apple Internal Keyboard  Trackpad: (accel) acceleration profile 0
[    87.874](**) Apple Internal Keyboard  Trackpad: (accel) acceleration factor: 2.000
[    87.874](**) Apple Internal Keyboard  Trackpad: (accel) acceleration threshold: 4
[    87.874](II) Apple Internal Keyboard  Trackpad: SetupAuto: hw.iftype is 4, hw.model is 0
[    87.874](II) Apple Internal Keyboard  Trackpad: SetupAuto: protocol is SysMouse
[    87.883](II) config/hal: Adding input device Apple Internal Keyboard  Trackpad
[    87.883](II) Using input driver 'kbd' for 'Apple Internal Keyboard  Trackpad'
[    87.883](**) Apple Internal Keyboard  Trackpad: always reports core events
[    87.883](**) Apple Internal Keyboard  Trackpad: always reports core events
[    87.883](**) Option "Protocol" "standard"
[    87.883](WW) Option "Device" requires an string value
[    87.883](**) Option "XkbRules" "base"
[    87.883](**) Option "XkbModel" "pc105"
[    87.883](**) Option "XkbLayout" "us"
[    87.883](**) Option "config_info" "hal:/org/freedesktop/Hal/devices/usb_device_5ac_21a_noserial_if0"

## [    87.883](II) XINPUT: Adding extended input device "Apple Internal Keyboard  Trackpad" (type: KEYBOARD, id 8)

And no worries, just thought I'd get it to you while the info was still warm. :)

**Comment by KevinMcA** [comment]
Once again, no rush, have compiled kernel with USB_DEBUG set, assuming that might help. Just wanted to let you know I'm all set for whenever you have a spare moment to help get this resolved. Enjoy, and thanks once again for offering! :)

**Comment by KevinMcA** [comment]
Hi! Since this is my day off, ran the hw.usb.atp.debug=19, and the results added up to about 1.9 megs worth. What might be of use is that as the sensitivity worked its way down the pad, the numbers from lower down towards the single button remained high as the sensitive area wound itself down to the bottom of the pad before nothing. In following with the fingers as it drifted downward, the new physical positions tended to yield the same results, so hope this helps. Be happy to forward the whole enchilada if you wish, but as the sensitivity sank down, I was getting the same numbers in different physical positions which suggests that somehow the top of the pad is not getting rest to the proper position. Here's one result to show you that when things were reacting at the top of the pad, we were getting the same data as that position continued to sink to the bottom of the pad until nothing:

lc:4500,dm:0,cum:19,max:35,],[lc:3349,dm:0,cum:27,max:42,] [lc:6236,dm:0,cum:17,max:30,],[lc:3349,dm:0,cum:27,max:39,]

lc ranged as high as 4800 when it was at the top, got similar as the pad sunk away down to the bottom and stopped producing anything. Let me know if you want to take this off-list and hopefully we can wipe all these additional messages off here. You can get to me at support at knosproject dot com.

**Comment by rgrover** [comment]
Hi,

Thanks for all this wonderful information. I'll start looking at it in this evening. I won't bother dusting off my old Mac; perhaps you can help collect the data (as you've been doing) and we can work on improving the interpretation algorithm together.

**Comment by KevinMcA** [comment]
Thanks for the kind words, I'd like to see this work too since it always 
has (aside from scrolling) since you originally wrote the code. It 
appears as though the roll-in with the wellspring stuff and making that 
fit somehow disturbed the position reset that always worked prior in 
that old code of yours.

Be absolutely happy to take this off-list so I can edit out so much of 
the unnecessary info for the masses to peek at, so here's my real email 
address. Headed to bed now, in my "real life" I work for Apple (don't be 
impressed, I'm just another phone droid At-home-advisor, heh) and 
they'll be eating my Monday and Tuesday, but will be ready to go again 
once I get granny's ipad hooking up to facebook again. :)

In the meantime, to clarify what's going on here so you have an easier 
"aha!" moment of your own, it looks like when zombies get reaped and the 
pad goes idle, instead of resetting to the physical top x-y coordinates, 
it's somehow incrementing Y until the code is slowly seeing the top move 
down and then off the bottom of the pad where no fingers are detected 
anymore.

I messed around with atp_set_device_mode() and atp_reset_callback() by 
setting sc->sc_mode_bytes[] to RAW_SENSOR_MODE since that's how it was 
handled originally, no joy there. With both functions set to that 
instead of 0x04, I _did_ end up with no more drift towards the bottom, 
but the finger movement was constrained to a ridiculously small box and 
not enough to do anything. But it DID stay in place though it didn't work.

Once that gave no joy, I came running for help. (grin)

Catch you Tuesday night if that works for you, I'm in upstate New York 
just so you know the UTC stuff.

THANKS once again!
